### PR TITLE
Update CMake files to support Windows platform

### PR DESCRIPTION
**Pull request by kaadam**
This PR aims to enable the build for Windows x64. Windows port of Flang is not ready yet, but it would be nice if anybody could run a build and check the current status.
This PR is created based on #464, since there is no any activity on that that's why I opened a new one.

### CODE REVIEW

**Comment by xoviat** [comment]
Note: I had planned to submit a PR to update the CMake files after the current PRs are merged. I'd like to wait until the runtime and pgmath PRs are merged, then I can see where we are.

**Comment by isuruf** [comment]
On windows if you are building with clang with MSVC ABI `MSVC` is still true. the only case where `WIN32` is true and `MSVC` is not is if you are using MinGW/MSYS. In that case, the libraries are more like unix with `libm` also available.

**Comment by kaadam** [comment]
> On windows if you are building with clang with MSVC ABI `MSVC` is still true. the only case where `WIN32` is true and `MSVC` is not is if you are using MinGW/MSYS. In that case, the libraries are more like unix with `libm` also available.

Yes, you're right. I tested with clang for windows. It seems It was a side effect of my built clang (I don't know why).  I dropped that change. Thanks for the comment.

**Comment by kaadam** [comment]
> Note: I had planned to submit a PR to update the CMake files after the current PRs are merged. I'd like to wait until the runtime and pgmath PRs are merged, then I can see where we are.

@xoviat sorry, I think it would be nice to enable the basic build, maybe it makes the review process easier.  In my opinion it could be merge independently of the other Windows changes, of course if you and others agree with that.

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,45 +9,52 @@ cmake_minimum_required(VERSION 3.3)
 # In order to bootstrap the runtime library we need to skip
 # CMake's Fortran tests
 SET(CMAKE_Fortran_COMPILER_WORKS 1)
+SET(CMAKE_Fortran_ABI_COMPILED 0)
+SET(CMAKE_Fortran_COMPILER_SUPPORTS_F90 1)
+SET(CMAKE_Fortran_MODDIR_FLAG "-J ")
```
Comment (bryanpkc) [inline CMakeLists.txt:14]: Do you run CMake with `-DCMAKE_Fortran_COMPILER_ID=Flang`? That should set `CMAKE_Fortran_MODDIR_FLAG` correctly.

File: CMakeLists.txt [hunk 2]
```diff
@@ -9,45 +9,52 @@ cmake_minimum_required(VERSION 3.3)
...
 
-if( NOT DEFINED TARGET_ARCHITECTURE )
-  execute_process(COMMAND uname -m OUTPUT_STRIP_TRAILING_WHITESPACE
-                  OUTPUT_VARIABLE TARGET_ARCHITECTURE)
+# If we are not building as a part of LLVM, build Flang as a
+# standalone project, using LLVM as an external library:
+if( CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR )
+  project(Flang)
```
Comment (bryanpkc) [inline CMakeLists.txt:19]: Why is this necessary? We already build Flang as a standalone project and haven't needed this.

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,45 +9,52 @@ cmake_minimum_required(VERSION 3.3)
...
 endif()
 
-if( NOT DEFINED TARGET_OS )
-  execute_process(COMMAND uname -s OUTPUT_STRIP_TRAILING_WHITESPACE
-                  OUTPUT_VARIABLE TARGET_OS)
+set(TARGET_OS ${CMAKE_SYSTEM_NAME} CACHE STRING "Target OS")
+set(TARGET_ARCHITECTURE ${CMAKE_SYSTEM_PROCESSOR} CACHE STRING "Target Architecture")
```
Comment (bryanpkc) [inline CMakeLists.txt:23]: Can you preserve the original convenient behaviour if `uname` can be found and is executable?

File: runtime/flang/CMakeLists.txt [hunk 4]
```diff
@@ -475,7 +479,15 @@ add_flang_library(flang_static
   ${FTN_SUPPORT_I8}
   ${SHARED_SOURCES}
   )
+set_target_properties(flang_static
+  PROPERTIES
+  Fortran_MODULE_DIRECTORY ${CMAKE_BINARY_DIR}/include-static
+)
+if (MSVC)
+set_property(TARGET flang_static PROPERTY OUTPUT_NAME libflang)
```
Comment (bryanpkc) [inline runtime/flang/CMakeLists.txt:487]: Please indent.

Comment (bryanpkc) [reply runtime/flang/CMakeLists.txt:487]: Line 489 as well.

**Comment by isuruf** [comment]
@bryanpkc, can you review #464 as that PR has the commits with authors correctly?

File: runtime/ompstub/CMakeLists.txt [hunk 5]
```diff
@@ -7,11 +7,18 @@
 set(OMPSTUB_SRC init_nomp.c  ompstubs.c)
 
 add_flang_library(ompstub_static ${OMPSTUB_SRC})
+if (MSVC)
+set_property(TARGET ompstub_static PROPERTY OUTPUT_NAME libompstub)
```
Comment (bryanpkc) [inline runtime/ompstub/CMakeLists.txt:11]: Please indent this line and line 13.

File: CMakeLists.txt (hunk 2 above)
Comment (bryanpkc) [reply CMakeLists.txt:19]: I see, thank you.
